### PR TITLE
Simplify semantic_tokens, fix off-by-one error in identifierFromPosition

### PR DESCRIPTION
**Pull request by InterplanetaryEngineer**
This should allow me to next remove the async recursion in a way that actually works.

### CODE REVIEW

File: src/offsets.zig [hunk 1]
```diff
@@ -80,7 +80,7 @@ pub fn lineSectionLength(tree: ast.Tree, start_index: usize, end_index: usize, e
 
         const codepoint = try std.unicode.utf8Decode(source[i .. i + n]);
 
-        result += 1 + @as(usize, @boolToInt(codepoint >= 0x10000));
+        result += 1 + @boolToInt(codepoint >= 0x10000);
```
Comment (alexnask) [inline src/offsets.zig:83]: The coercion here is intentional, @boolToInt will return `u1` and thus the whole expression is of type `u1` and will overflow when the condition is true.

**Review by alexnask — CHANGES_REQUESTED** [review]
The rest of this code looks fine  
I will be testing this in the morning, thanks for the contribution

**Comment by alexnask** [comment]
The CI started erroring out with BrokenPipe while running the session tests a couple of days ago, I will investigate and fix this.

File: src/offsets.zig (hunk 1 above)
Comment (alexnask) [reply src/offsets.zig]: Maybe
```zig
result = result + 1 + @boolToInt(codepoint >= 0x10000);
```
is saner here for this reason

Comment (InterplanetaryEngineer) [reply src/offsets.zig]: Thanks for the catch, fixed it.

**Review by alexnask — APPROVED** [review]

**Comment by alexnask** [comment]
Looks good, thanks again.

**Pull request merged**